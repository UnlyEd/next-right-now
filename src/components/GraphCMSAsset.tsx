import classnames from 'classnames';
import deepmerge from 'deepmerge';
import get from 'lodash.get';
import isEmpty from 'lodash.isempty';
import map from 'lodash.map';
import PropTypes from 'prop-types';
import React from 'react';
import stylePropType from 'react-style-proptype';

import GraphCMSAssetPropTypes from '../propTypes/GraphCMSAssetPropTypes';
import GraphCMSAssetTransformationsPropTypes from '../propTypes/GraphCMSAssetTransformationsPropTypes';
import { Asset } from '../types/data/Asset';
import { AssetTransformations } from '../types/data/AssetTransformations';
import { Link } from '../types/data/Link';
import { cssToReactStyle } from '../utils/css';

const _defaultAsset = {
  id: null,
  url: '', // TODO use link of default missing image
  title: '',
  alt: '',
  style: null,
  className: '',
};

const _defaultLink: Link = {
  id: null,
  url: null,
  target: '_blank',
  style: null,
  className: '',
};

/**
 * Displays an asset, based on the provided props
 * Handles GraphCMS assets, including svg
 * Should be used to display assets coming from GraphCMS (handles transformations)
 *
 * @param props
 * @return {null|*}
 * @constructor
 *
 * @see Transformations https://docs.graphcms.com/developers/assets/transformations/transforming-url-structure
 */
const GraphCMSAsset = (props: Props): JSX.Element => {
  const {
    id,
    asset,
    defaults = {},
    override = {},
    className = '',
    style = null,
    onClick = null,
    linkOverride = {},
    transformationsOverride = null,
    forcePNGOutput = true,
  }: Props = props;
  if (isEmpty(asset)) {
    return null;
  }
  const resolvedAssetProps: Asset = deepmerge.all([_defaultAsset, defaults, asset || {}, override]);
  const resolvedLinkProps: Link = deepmerge.all([
    _defaultLink,
    {
      url: asset.linkUrl || _defaultLink.url,
      target: asset.linkTarget || _defaultLink.target,
    },
    linkOverride,
  ]);
  const transformations: AssetTransformations = transformationsOverride || asset.defaultTransformations;

  // Convert "style" if it is a string, to a react style object (won't modify if already an object)
  resolvedAssetProps.style = cssToReactStyle(resolvedAssetProps.style);
  resolvedLinkProps.style = cssToReactStyle(resolvedLinkProps.style);

  if (transformations) {
    // XXX Transformations are handled differently when the file is an SVG
    if (get(asset, 'mimeType') === 'image/svg+xml') {
      // If the asset is a SVG, then we must set width/height manually as styles, because transformations won't work on this kind of asset
      if (transformations.width) {
        resolvedAssetProps.style.width = transformations.width;
      }
      if (transformations.height) {
        resolvedAssetProps.style.height = transformations.height;
      }

    } else {
      const _urlSplitted = resolvedAssetProps.url.split('/'); // TODO will break if url ends with '/'
      const assetFileHandle = _urlSplitted[_urlSplitted.length - 1];
      const assetBaseUrl = resolvedAssetProps.url.replace(assetFileHandle, '');

      // XXX See https://graphcms.com/docs/assets/transformations/
      let transformationsToApply = 'quality=value:100/'; // Forcing max quality (not sure it's useful, but we want max visual quality)
      transformationsToApply += forcePNGOutput ? 'output=format:png/' : ''; // Forcing output format to be PNG, allows to display PDF files as images (this bypasses the usage of SVG files)

      // Handle resize transformations
      const resize: AssetTransformations = {};
      if (transformations.width) {
        resize.width = transformations.width;
      }
      if (transformations.height) {
        resize.height = transformations.height;
      }

      if (!isEmpty(resize)) {
        transformationsToApply += 'resize=' + map(resize, (value: string, transformation: string) => {
          return transformation + ':' + value;
        }).join(',');
      }

      // Once all transformations have been resolved, update the asset url
      resolvedAssetProps.url = `${assetBaseUrl}${transformationsToApply}/${assetFileHandle}`;
    }
  }

  const Image = (): JSX.Element => {
    return (
      <img
        key={id}
        id={id}
        src={resolvedAssetProps.url}
        title={resolvedAssetProps.title}
        alt={resolvedAssetProps.alt}
        className={classnames(`asset-${id}`, className, resolvedAssetProps.classes)}
        style={deepmerge(style || {}, resolvedAssetProps.style || {})}
      />
    );
  };

  // Create a link wrapper only if a link url is provided
  if (resolvedLinkProps.url) {
    return (
      <a
        key={resolvedLinkProps.id}
        id={resolvedLinkProps.id}
        href={resolvedLinkProps.url}
        target={resolvedLinkProps.target}
        className={classnames(`asset-link-${id}`, resolvedLinkProps.classes, resolvedLinkProps.className)}
        style={deepmerge(resolvedLinkProps.style || {}, resolvedLinkProps.style || {})}
        // @ts-ignore
        onClick={onClick} // Support for usage within <Link> component (from Next.js)
      >
        <Image />
      </a>
    );
  } else {
    return (
      <Image />
    );
  }
};

GraphCMSAsset.propTypes = {
  id: PropTypes.string.isRequired,
  asset: PropTypes.shape(GraphCMSAssetPropTypes).isRequired,
  transformationsOverride: PropTypes.shape(GraphCMSAssetTransformationsPropTypes),
  defaults: PropTypes.shape(GraphCMSAssetPropTypes), // Merged with the asset, takes lowest priority
  override: PropTypes.shape(GraphCMSAssetPropTypes), // Merged with the asset, takes highest priority
  className: PropTypes.string,
  style: stylePropType,
  onClick: PropTypes.func, // Support for usage within <Link> component (from Next.js)
  linkOverride: PropTypes.shape({ // Merged with the default link and the asset link attributes, takes highest priority
    id: PropTypes.string.isRequired,
    url: PropTypes.string,
    target: PropTypes.string,
    style: PropTypes.object,
    classes: PropTypes.string,
  }),
  forceOutputPNG: PropTypes.bool,
};

type Props = {
  id: string;
  asset: Asset;
  transformationsOverride?: AssetTransformations;
  defaults?: Asset;
  override?: Asset;
  className?: string;
  style?: object;
  onClick?: Function;
  linkOverride?: {
    id?: string;
    url?: string;
    target?: string;
    style?: object;
    classes?: string;
  };
  forcePNGOutput?: boolean;
}

export default GraphCMSAsset;
