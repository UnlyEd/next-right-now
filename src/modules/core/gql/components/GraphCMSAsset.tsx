import { cssToReactStyle } from '@/modules/core/css/css';
import { CSSStyles } from '@/modules/core/css/types/CSSStyles';
import { Asset } from '@/modules/core/data/types/Asset';
import { AssetTransformations } from '@/modules/core/data/types/AssetTransformations';
import classnames from 'classnames';
import deepmerge from 'deepmerge';
import get from 'lodash.get';
import isEmpty from 'lodash.isempty';
import map from 'lodash.map';
import React from 'react';
import { Link } from '../../data/types/Link';

type Props = {
  /**
   * Asset, extends GraphCMS asset, e.g: image, document, etc.
   *
   * <span className="tip">XXX</span> The `asset` will be merged with `_defaultAsset`.
   */
  asset: Asset;

  /**
   * HTML id attribute. Must be unique.
   */
  id: string;

  /**
   * Overrides transformations.
   *
   * @default null
   */
  transformationsOverride?: AssetTransformations;

  /**
   * Default `asset` properties.
   *
   * @default {}
   */
  defaults?: Asset;

  /**
   * Overrides `asset` properties.
   *
   * @default {}
   */
  override?: Asset;

  /**
   * CSS classes.
   */
  className?: string;

  /**
   * CSS styles
   *
   * @default null
   */
  style?: CSSStyles;

  /**
   * Click event.
   */
  onClick?: () => void;

  /**
   * Overrides `Link` element properties.
   *
   * @default {}
   */
  linkOverride?: {
    id?: string;
    url?: string;
    target?: string;
    style?: CSSStyles;
    classes?: string;
  };

  /**
   * Whether to force output format to be PNG, allows to display PDF files as images.
   *
   * It bypasses the usage of SVG files.
   *
   * @default false
   */
  forcePNGOutput?: boolean;
}

const _defaultAsset = {
  id: null,
  url: '', // TODO use link of default missing image
  title: '',
  alt: '',
  style: null,
  className: '',
};

const _defaultLink: Link = {
  id: null,
  url: null,
  target: '_blank',
  style: null,
  className: '',
};

/**
 * Displays an asset, based on the provided props
 * Handles GraphCMS assets, including svg
 * Should be used to display assets coming from GraphCMS (handles transformations)
 *
 * @param props
 * @return {null|*}
 *
 * @see Transformations https://docs.graphcms.com/developers/assets/transformations/transforming-url-structure
 */
const GraphCMSAsset: React.FunctionComponent<Props> = (props): JSX.Element => {
  const {
    id,
    asset,
    defaults = {},
    override = {},
    className = '',
    style = null,
    onClick = null,
    linkOverride = {},
    transformationsOverride = null,
    forcePNGOutput = false,
  }: Props = props;
  if (isEmpty(asset)) {
    return null;
  }
  const resolvedAssetProps: Asset = deepmerge.all([_defaultAsset, defaults, asset || {}, override]);
  const resolvedLinkProps: Link = deepmerge.all([
    _defaultLink,
    {
      url: asset.linkUrl || _defaultLink.url,
      target: asset.linkTarget || _defaultLink.target,
    },
    linkOverride,
  ]);
  const transformations: AssetTransformations = transformationsOverride || asset.defaultTransformations;

  // Convert "style" if it is a string, to a react style object (won't modify if already an object)
  resolvedAssetProps.style = cssToReactStyle(resolvedAssetProps.style);
  resolvedLinkProps.style = cssToReactStyle(resolvedLinkProps.style);

  if (transformations) {
    // XXX Transformations are handled differently when the file is an SVG
    if (get(asset, 'mimeType') === 'image/svg+xml') {
      // If the asset is a SVG, then we must set width/height manually as styles, because transformations won't work on this kind of asset
      if (transformations.width) {
        resolvedAssetProps.style.width = transformations.width;
      }
      if (transformations.height) {
        resolvedAssetProps.style.height = transformations.height;
      }

    } else {
      const _urlSplitted = resolvedAssetProps.url.split('/'); // TODO will break if url ends with '/'
      const assetFileHandle = _urlSplitted[_urlSplitted.length - 1];
      const assetBaseUrl = resolvedAssetProps.url.replace(assetFileHandle, '');

      // XXX See https://docs.graphcms.com/docs/assets/transformations/
      let transformationsToApply = 'quality=value:100/'; // Forcing max quality (not sure it's useful, but we want max visual quality)
      transformationsToApply += forcePNGOutput ? 'output=format:png/' : ''; // Forcing output format to be PNG, allows to display PDF files as images (this bypasses the usage of SVG files)

      // Handle resize transformations
      const resize: AssetTransformations = {};
      if (transformations.width) {
        resize.width = transformations.width;
      }
      if (transformations.height) {
        resize.height = transformations.height;
      }

      if (!isEmpty(resize)) {
        transformationsToApply += 'resize=' + map(resize, (value: string, transformation: string) => {
          return transformation + ':' + value;
        }).join(',');
      }

      // Once all transformations have been resolved, update the asset url
      // XXX Using "auto_image" will automatically select the image type (WebP, etc.) based on the browser (reduces size by 20-80%, compared to png)
      //  See https://www.filestack.com/docs/api/processing/#auto-image-conversion
      resolvedAssetProps.url = `${assetBaseUrl}${transformationsToApply}/auto_image/${assetFileHandle}`;
    }
  }

  const Image = (): JSX.Element => {
    return (
      // eslint-disable-next-line @next/next/no-img-element
      <img
        key={id}
        id={id}
        src={resolvedAssetProps.url}
        title={resolvedAssetProps.title}
        alt={resolvedAssetProps.alt || resolvedAssetProps.title || resolvedAssetProps.url}
        className={classnames(`asset-${id}`, className, resolvedAssetProps.classes)}
        style={deepmerge(style || {}, resolvedAssetProps.style || {})}
      />
    );
  };

  // Create a link wrapper only if a link url is provided
  if (resolvedLinkProps.url) {
    return (
      <a
        key={resolvedLinkProps.id}
        id={resolvedLinkProps.id}
        href={resolvedLinkProps.url}
        target={resolvedLinkProps.target}
        className={classnames(`asset-link-${id}`, resolvedLinkProps.classes, resolvedLinkProps.className)}
        style={deepmerge(resolvedLinkProps.style || {}, resolvedLinkProps.style || {})}
        onClick={onClick} // Support for usage within <Link> component (from Next.js)
      >
        <Image />
      </a>
    );
  } else {
    return (
      <Image />
    );
  }
};

export default GraphCMSAsset;
