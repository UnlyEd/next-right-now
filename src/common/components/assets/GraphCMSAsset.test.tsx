import React from 'react';
import TestRenderer from 'react-test-renderer';
import GraphCMSAsset from './GraphCMSAsset';

const defaultLogoUrl = 'https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V';
const defaultLogoTarget = '_blank';

describe('GraphCMSAsset', () => {
  describe('should properly render an asset from GraphCMS', () => {
    describe('when the asset is used as an image (<img>)', () => {
      test('when relying on default "logo" property, it should apply the internal default properties', () => {
        const id = 'test';
        const renderer = TestRenderer
          .create(<GraphCMSAsset
            id={id}
            asset={{
              url: defaultLogoUrl,
            }}
          />);
        const img = renderer.toJSON();

        expect(img.props.id).toEqual(id);
        expect(img.props.src).toEqual(defaultLogoUrl);
        expect(img.props.title).toEqual('');
        expect(img.props.alt).toEqual('');
        expect(img.props.className).toEqual(`asset-${id}`);
        expect(img.props.style).toEqual({});
        expect(img).toMatchSnapshot();
      });

      test('when using "logo" property, it should apply the properties correctly', () => {
        const id = 'test';
        const title = 'Test asset';
        const classes = 'test-class1 test-class2';
        const style = { paddingTop: 5 };
        const renderer = TestRenderer
          .create(<GraphCMSAsset
            id={id}
            asset={{
              url: defaultLogoUrl,
              title: title,
              alt: title,
              classes: classes,
              style: style,
            }}
          />);
        const img = renderer.toJSON();

        expect(img.props.id).toEqual(id);
        expect(img.props.src).toEqual(defaultLogoUrl);
        expect(img.props.title).toEqual(title);
        expect(img.props.alt).toEqual(title);
        expect(img.props.className).toEqual(`asset-${id} ${classes}`);
        expect(img.props.style).toEqual(style);
        expect(img).toMatchSnapshot();
      });

      describe('when using custom defaults', () => {
        test('which should be properly applied', () => {
          const id = 'test';
          const title = 'Test asset';
          const classes = 'test-class1 test-class2';
          const style = { paddingTop: 5 };
          const renderer = TestRenderer
            .create(<GraphCMSAsset
              id={id}
              asset={{
                url: defaultLogoUrl,
                title: title,
                alt: title,
              }}
              defaults={{
                classes: classes,
                style: style,
              }}
            />);
          const img = renderer.toJSON();

          expect(img.props.id).toEqual(id);
          expect(img.props.src).toEqual(defaultLogoUrl);
          expect(img.props.title).toEqual(title);
          expect(img.props.alt).toEqual(title);
          expect(img.props.className).toEqual(`asset-${id} ${classes}`);
          expect(img.props.style).toEqual(style);
          expect(img).toMatchSnapshot();
        });
      });

      describe('when using custom override', () => {
        test('when used with the "logo" property, it should override the logo props with the override values', () => {
          const id = 'test';
          const title = 'Test asset';
          const classes = 'test-class1 test-class2';
          const style = { paddingTop: 5 };
          const renderer = TestRenderer
            .create(<GraphCMSAsset
              id={id}
              asset={{
                url: defaultLogoUrl,
                title: title,
                alt: title,
              }}
              override={{
                classes: classes,
                style: style,
              }}
            />);
          const img = renderer.toJSON();

          expect(img.props.id).toEqual(id);
          expect(img.props.src).toEqual(defaultLogoUrl);
          expect(img.props.title).toEqual(title);
          expect(img.props.alt).toEqual(title);
          expect(img.props.className).toEqual(`asset-${id} ${classes}`);
          expect(img.props.style).toEqual(style);
          expect(img).toMatchSnapshot();
        });

        test('when used with the "defaults" property, it should override defaults and asset properties with the overridden values', () => {
          const id = 'test';
          const title = 'Test asset';
          const classes = 'test-class1 test-class2';
          const style = { paddingTop: 5 };
          const renderer = TestRenderer
            .create(<GraphCMSAsset
              id={id}
              asset={{
                url: defaultLogoUrl,
                title: title,
                alt: title,
                classes: 'should-not-be-applied',
                style: { paddingTop: 1 }, // Should be overridden
              }}
              defaults={{
                classes: 'should-not-be-applied2',
                style: { paddingTop: 3 }, // Should be overridden
              }}
              override={{
                classes: classes,
                style: style,
              }}
            />);
          const img = renderer.toJSON();

          expect(img.props.id).toEqual(id);
          expect(img.props.src).toEqual(defaultLogoUrl);
          expect(img.props.title).toEqual(title);
          expect(img.props.alt).toEqual(title);
          expect(img.props.className).toEqual(`asset-${id} ${classes}`);
          expect(img.props.style).toEqual(style);
          expect(img).toMatchSnapshot();
        });
      });

      test('when using multiple "transformations" properties, it should apply all transformations', () => {
        const id = 'test';
        const title = 'Test asset';
        const classes = 'test-class1 test-class2';
        const style = { paddingTop: 5 };
        const renderer = TestRenderer
          .create(<GraphCMSAsset
            id={id}
            asset={{
              url: defaultLogoUrl,
              title: title,
              alt: title,
              classes: classes,
              style: style,
              defaultTransformations: {
                width: 500,
                height: 300,
              },
            }}
          />);
        const img = renderer.toJSON();

        expect(img.props.id).toEqual(id);
        expect(img.props.src).toEqual('https://media.graphcms.com/quality=value:100/output=format:png/resize=width:500,height:300/88YmsSFsSEGI9i0qcH0V');
        expect(img.props.title).toEqual(title);
        expect(img.props.alt).toEqual(title);
        expect(img.props.className).toEqual(`asset-${id} ${classes}`);
        expect(img.props.style).toEqual(style);
        expect(img).toMatchSnapshot();
      });

      test('when using a single "transformations" property, it should apply this transformation', () => {
        const id = 'test';
        const title = 'Test asset';
        const classes = 'test-class1 test-class2';
        const style = { paddingTop: 5 };
        const renderer = TestRenderer
          .create(<GraphCMSAsset
            id={id}
            asset={{
              url: defaultLogoUrl,
              title: title,
              alt: title,
              classes: classes,
              style: style,
              defaultTransformations: {
                width: 500,
              },
            }}
          />);
        const img = renderer.toJSON();

        expect(img.props.id).toEqual(id);
        expect(img.props.src).toEqual('https://media.graphcms.com/quality=value:100/output=format:png/resize=width:500/88YmsSFsSEGI9i0qcH0V');
        expect(img.props.title).toEqual(title);
        expect(img.props.alt).toEqual(title);
        expect(img.props.className).toEqual(`asset-${id} ${classes}`);
        expect(img.props.style).toEqual(style);
        expect(img).toMatchSnapshot();
      });

      test('when using "transformationsOverride" property, it should only use the override and ignore the default transformations', () => {
        const id = 'test';
        const title = 'Test asset';
        const classes = 'test-class1 test-class2';
        const style = { paddingTop: 5 };
        const renderer = TestRenderer
          .create(<GraphCMSAsset
            id={id}
            asset={{
              url: defaultLogoUrl,
              title: title,
              alt: title,
              classes: classes,
              style: style,
              defaultTransformations: {
                width: 500,
              },
            }}
            transformationsOverride={{
              height: 300
            }}
          />);
        const img = renderer.toJSON();

        expect(img.props.id).toEqual(id);
        expect(img.props.src).toEqual('https://media.graphcms.com/quality=value:100/output=format:png/resize=height:300/88YmsSFsSEGI9i0qcH0V');
        expect(img.props.title).toEqual(title);
        expect(img.props.alt).toEqual(title);
        expect(img.props.className).toEqual(`asset-${id} ${classes}`);
        expect(img.props.style).toEqual(style);
        expect(img).toMatchSnapshot();
      });
    });
  });

  describe('when the asset is used as a link(<a>) containing an image (<img>)', () => {
    test('when relying on default "logo" property', () => {
      const id = 'test';
      const linkUrl = 'https://google.com';
      const renderer = TestRenderer
        .create(<GraphCMSAsset
          id={id}
          asset={{
            url: defaultLogoUrl,
            linkUrl: linkUrl,
          }}
        />);
      const link = renderer.toJSON();
      const img = renderer.root.findByType('img');

      expect(img.props.id).toEqual(id);
      expect(img.props.src).toEqual(defaultLogoUrl);
      expect(img.props.title).toEqual('');
      expect(img.props.alt).toEqual('');
      expect(img.props.className).toEqual(`asset-${id}`);
      expect(img.props.style).toEqual({});
      expect(img.props).toMatchSnapshot();

      expect(link.props.id).toEqual(null);
      expect(link.props.href).toEqual(linkUrl);
      expect(link.props.target).toEqual('_blank');
      expect(link).toMatchSnapshot();
    });

    test('when using "linkOverride" property', () => {
      const id = 'test';
      const title = 'Test asset';
      const classes = 'test-class1 test-class2';
      const style = { paddingTop: 5 };
      const linkId = 'link-test';
      const linkUrl = 'https://google.com';

      const renderer = TestRenderer
        .create(<GraphCMSAsset
          id={id}
          asset={{
            url: defaultLogoUrl,
            title: title,
            alt: title,
            classes: classes,
            style: style,
            linkUrl: linkUrl,
          }}
          linkOverride={{
            id: linkId,
            classes: classes,
            style: style,
          }}
        />);
      const link = renderer.toJSON();
      const img = renderer.root.findByType('img');

      expect(img.props.id).toEqual(id);
      expect(img.props.src).toEqual(defaultLogoUrl);
      expect(img.props.title).toEqual(title);
      expect(img.props.alt).toEqual(title);
      expect(img.props.className).toEqual(`asset-${id} ${classes}`);
      expect(img.props.style).toEqual(style);
      expect(img.props).toMatchSnapshot();

      expect(link.props.id).toEqual(linkId);
      expect(link.props.href).toEqual(linkUrl);
      expect(link.props.target).toEqual('_blank');
      expect(link.props.className).toEqual(`asset-link-${id} ${classes}`);
      expect(link.props.style).toEqual(style);
      expect(link).toMatchSnapshot();
    });

    test('when using "linkOverride" property to override the link url', () => {
      const id = 'test';
      const title = 'Test asset';
      const classes = 'test-class1 test-class2';
      const style = { paddingTop: 5 };
      const linkId = 'link-test';
      const linkUrl = 'https://google.com';
      const overriddenLinkUrl = 'https://overridden.com';

      const renderer = TestRenderer
        .create(<GraphCMSAsset
          id={id}
          asset={{
            url: defaultLogoUrl,
            title: title,
            alt: title,
            classes: classes,
            style: style,
            linkUrl: linkUrl,
          }}
          linkOverride={{
            id: linkId,
            url: overriddenLinkUrl,
            classes: classes,
            style: style,
          }}
        />);
      const link = renderer.toJSON();
      const img = renderer.root.findByType('img');

      expect(img.props.id).toEqual(id);
      expect(img.props.src).toEqual(defaultLogoUrl);
      expect(img.props.title).toEqual(title);
      expect(img.props.alt).toEqual(title);
      expect(img.props.className).toEqual(`asset-${id} ${classes}`);
      expect(img.props.style).toEqual(style);
      expect(img.props).toMatchSnapshot();

      expect(link.props.id).toEqual(linkId);
      expect(link.props.href).toEqual(overriddenLinkUrl);
      expect(link.props.target).toEqual('_blank');
      expect(link.props.className).toEqual(`asset-link-${id} ${classes}`);
      expect(link.props.style).toEqual(style);
      expect(link).toMatchSnapshot();
    });

    test('when using "onClick" property', () => {
      const id = 'test';
      const title = 'Test asset';
      const classes = 'test-class1 test-class2';
      const style = { paddingTop: 5 };
      const linkId = 'link-test';
      const linkUrl = 'https://google.com';
      const overriddenLinkUrl = 'https://overridden.com';
      const onClick = () => 'clicked';

      const renderer = TestRenderer
        .create(<GraphCMSAsset
          id={id}
          asset={{
            url: defaultLogoUrl,
            title: title,
            alt: title,
            classes: classes,
            style: style,
            linkUrl: linkUrl,
          }}
          linkOverride={{
            id: linkId,
            url: overriddenLinkUrl,
            classes: classes,
            style: style,
          }}
          onClick={onClick}
        />);
      const link = renderer.toJSON();
      const img = renderer.root.findByType('img');

      expect(img.props.id).toEqual(id);
      expect(img.props.src).toEqual(defaultLogoUrl);
      expect(img.props.title).toEqual(title);
      expect(img.props.alt).toEqual(title);
      expect(img.props.className).toEqual(`asset-${id} ${classes}`);
      expect(img.props.style).toEqual(style);
      expect(img.props).toMatchSnapshot();

      expect(link.props.id).toEqual(linkId);
      expect(link.props.href).toEqual(overriddenLinkUrl);
      expect(link.props.target).toEqual('_blank');
      expect(link.props.className).toEqual(`asset-link-${id} ${classes}`);
      expect(link.props.style).toEqual(style);
      expect(link.props.onClick).toBeInstanceOf(Function);
      expect(link.props.onClick()).toEqual('clicked');
      expect(link).toMatchSnapshot();
    });

  });
});
