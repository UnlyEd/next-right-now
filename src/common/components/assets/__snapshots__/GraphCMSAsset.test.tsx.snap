// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when relying on default "logo" property, it should apply the internal default properties 1`] = `
<img
  alt=""
  className="asset-test"
  id="test"
  src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
  style={Object {}}
  title=""
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using "logo" property, it should apply the properties correctly 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using "transformationsOverride" property, it should only use the override and ignore the default transformations 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/quality=value:100/output=format:png/resize=height:300/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using a single "transformations" property, it should apply this transformation 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/quality=value:100/output=format:png/resize=width:500/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using custom defaults which should be properly applied 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using custom override when used with the "defaults" property, it should override defaults and asset properties with the overridden values 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using custom override when used with the "logo" property, it should override the logo props with the override values 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset should properly render an asset from GraphCMS when the asset is used as an image (<img>) when using multiple "transformations" properties, it should apply all transformations 1`] = `
<img
  alt="Test asset"
  className="asset-test test-class1 test-class2"
  id="test"
  src="https://media.graphcms.com/quality=value:100/output=format:png/resize=width:500,height:300/88YmsSFsSEGI9i0qcH0V"
  style={
    Object {
      "paddingTop": 5,
    }
  }
  title="Test asset"
/>
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when relying on default "logo" property 1`] = `
Object {
  "alt": "",
  "className": "asset-test",
  "id": "test",
  "src": "https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V",
  "style": Object {},
  "title": "",
}
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when relying on default "logo" property 2`] = `
<a
  className="asset-link-test"
  href="https://google.com"
  id={null}
  onClick={null}
  style={Object {}}
  target="_blank"
>
  <img
    alt=""
    className="asset-test"
    id="test"
    src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
    style={Object {}}
    title=""
  />
</a>
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when using "linkOverride" property 1`] = `
Object {
  "alt": "Test asset",
  "className": "asset-test test-class1 test-class2",
  "id": "test",
  "src": "https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V",
  "style": Object {
    "paddingTop": 5,
  },
  "title": "Test asset",
}
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when using "linkOverride" property 2`] = `
<a
  className="asset-link-test test-class1 test-class2"
  href="https://google.com"
  id="link-test"
  onClick={null}
  style={
    Object {
      "paddingTop": 5,
    }
  }
  target="_blank"
>
  <img
    alt="Test asset"
    className="asset-test test-class1 test-class2"
    id="test"
    src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
    style={
      Object {
        "paddingTop": 5,
      }
    }
    title="Test asset"
  />
</a>
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when using "linkOverride" property to override the link url 1`] = `
Object {
  "alt": "Test asset",
  "className": "asset-test test-class1 test-class2",
  "id": "test",
  "src": "https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V",
  "style": Object {
    "paddingTop": 5,
  },
  "title": "Test asset",
}
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when using "linkOverride" property to override the link url 2`] = `
<a
  className="asset-link-test test-class1 test-class2"
  href="https://overridden.com"
  id="link-test"
  onClick={null}
  style={
    Object {
      "paddingTop": 5,
    }
  }
  target="_blank"
>
  <img
    alt="Test asset"
    className="asset-test test-class1 test-class2"
    id="test"
    src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
    style={
      Object {
        "paddingTop": 5,
      }
    }
    title="Test asset"
  />
</a>
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when using "onClick" property 1`] = `
Object {
  "alt": "Test asset",
  "className": "asset-test test-class1 test-class2",
  "id": "test",
  "src": "https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V",
  "style": Object {
    "paddingTop": 5,
  },
  "title": "Test asset",
}
`;

exports[`GraphCMSAsset when the asset is used as a link(<a>) containing an image (<img>) when using "onClick" property 2`] = `
<a
  className="asset-link-test test-class1 test-class2"
  href="https://overridden.com"
  id="link-test"
  onClick={[Function]}
  style={
    Object {
      "paddingTop": 5,
    }
  }
  target="_blank"
>
  <img
    alt="Test asset"
    className="asset-test test-class1 test-class2"
    id="test"
    src="https://media.graphcms.com/88YmsSFsSEGI9i0qcH0V"
    style={
      Object {
        "paddingTop": 5,
      }
    }
    title="Test asset"
  />
</a>
`;
